cmd/kubectl-gadget/opensnoop.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/kinvolk/inspektor-gadget/cmd/kubectl-gadget/utils"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/opensnoop/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+	"github.com/spf13/cobra"
+)
+
+var opensnoopCmd = &cobra.Command{
+	Use:   "opensnoop",
+	Short: "Trace open() system calls",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		var transform func(string) string
+		switch params.OutputMode {
+		case utils.OutputModeJson: // don't print any header
+		case utils.OutputModeCustomColumns:
+			fmt.Println(getCustomopensnoopColsHeader(params.CustomColumns))
+			transform = formatEventopensnoopCostumCols
+		case utils.OutputModeColumns:
+			transform = opensnoopTransformLine
+			fmt.Printf("%-16s %-16s %-16s %-16s %-6s %-16s %-3s %3s %s\n",
+				"NODE", "NAMESPACE", "POD", "CONTAINER",
+				"PID", "COMM", "FD", "ERR", "PATH")
+		}
+
+		config := &utils.TraceConfig{
+			GadgetName:       "opensnoop",
+			Operation:        "start",
+			TraceOutputMode:  "Stream",
+			TraceOutputState: "Started",
+			CommonFlags:      &params,
+		}
+
+		err := utils.RunTraceAndPrintStream(config, transform)
+		if err != nil {
+			return fmt.Errorf("failed to run tracer: %w", err)
+		}
+
+		return nil
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(opensnoopCmd)
+	utils.AddCommonFlags(opensnoopCmd, &params)
+}
+
+func opensnoopTransformLine(line string) string {
+	var sb strings.Builder
+	var e types.Event
+
+	if err := json.Unmarshal([]byte(line), &e); err != nil {
+		fmt.Fprintf(os.Stderr, "error unmarshalling json: %s", err)
+		return ""
+	}
+
+	switch e.Type {
+	case eventtypes.NORMAL:
+		sb.WriteString(fmt.Sprintf("%-16s %-16s %-16s %-16s %-6d %-16s %-3d %3d %s",
+			e.Node, e.Namespace, e.Pod, e.Container,
+			e.Pid, e.Comm, e.Fd, e.Err, e.Path))
+
+		return sb.String()
+	case eventtypes.ERR:
+		fmt.Fprintf(os.Stderr, "error from node %s: %s", e.Node, e.Message)
+		return ""
+	}
+
+	return ""
+}
+
+func getCustomopensnoopColsHeader(cols []string) string {
```

Comment:
Do you think these functions can be factorized and for example only giving the specific part inside a lambda?

Reply:
I think there still some room for improving it, but I'm not sure how to do it. Perhaps we can use a map with that data in to avoid writing all those lines :thinking: